builtin/credential/approle/path_tidy_user_id.go

```diff
@@ -39,206 +40,224 @@ func (b *backend) tidySecretID(ctx context.Context, req *logical.Request) (*logi
 		return resp, nil
 	}
 
-	s := req.Storage
+	go b.tidySecretIDbackground(req.Storage)
 
-	go func() {
-		defer atomic.StoreUint32(b.tidySecretIDCASGuard, 0)
+	resp := &logical.Response{}
+	resp.AddWarning("Tidy operation successfully started. Any information from the operation will be printed to Vault's server logs.")
+	return logical.RespondWithStatusCode(resp, req, http.StatusAccepted)
 
-		logger := b.Logger().Named("tidy")
+}
 
-		checkCount := 0
+type tidyHelperSecretIDAccessor struct {
+	secretIDAccessorStorageEntry
+	saltedSecretIDAccessor string
+}
 
-		defer func() {
-			if b.testTidyDelay > 0 {
-				logger.Trace("done checking entries", "num_entries", checkCount)
-			}
-		}()
+func (b *backend) tidySecretIDbackground(s logical.Storage) {
+	defer atomic.StoreUint32(b.tidySecretIDCASGuard, 0)
 
-		// Don't cancel when the original client request goes away
-		ctx = context.Background()
+	logger := b.Logger().Named("tidy")
 
-		tidyFunc := func(secretIDPrefixToUse, accessorIDPrefixToUse string) error {
-			logger.Trace("listing role HMACs", "prefix", secretIDPrefixToUse)
+	checkCount := 0
 
-			roleNameHMACs, err := s.List(ctx, secretIDPrefixToUse)
-			if err != nil {
-				return err
-			}
+	defer func() {
+		logger.Trace("done checking entries", "num_entries", checkCount)
+	}()
 
-			logger.Trace("listing accessors", "prefix", accessorIDPrefixToUse)
+	// Don't cancel when the original client request goes away
+	ctx := context.Background()
 
-			// List all the accessors and add them all to a map
-			accessorHashes, err := s.List(ctx, accessorIDPrefixToUse)
+	salt, err := b.Salt(ctx)
+	if err != nil {
+		logger.Error("error tidying secret IDs", "error", err)
+		return
+	}
+
+	tidyFunc := func(secretIDPrefixToUse, accessorIDPrefixToUse string) error {
+		logger.Trace("listing role HMACs", "prefix", secretIDPrefixToUse)
+
+		roleNameHMACs, err := s.List(ctx, secretIDPrefixToUse)
```

Comment:
Can we move this down to the line just above using `roleNameHMACs`?

Reply:
Will do.